keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1537 @@
+support code will be made available to simplify the development of
+such a plugin, but using it will not be required and its API is not
+part of this KEP.
+
+Three new API object types get added in a new API group:
+- ResourcePlugin, not namespaced, with a description of the plugin.
+- ResourceClass, not namespaced, with privileged parameters for
+  multiple resource instances of a certain kind. All these instances
+  are provided by the same resource plugin, which is identified by a
+  field in the class.
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual meta data, a spec and a status. The
+  spec identifies the plugin that handles the resource via a class
+  name.
+
+To support arbitrarily complex parameters, both ResourceClass and
+ResourceClaim contain one field which holds a
+runtime.RawExtension. Validation can be handled by plugins through an
+admission controller (if desired) or later at runtime when the
+parameters are passed to the plugin.
+
+The ResourceClaim spec is read-only once created. The ResourceClaim
+status is reserved for system usage and holds the current state of the
+resource. The status must not get lost. This is departing from how
+Kubernetes traditionally handled status, but something that more
+recently [became more
+acceptable](https://github.com/kubernetes/enhancements/pull/2537). Kube-scheduler
+and plugin communicate by modifying that status. The status is also
+how Kubernetes tracks that a plugin has allocated the resource and on
+which nodes the resource is available.
+
+This approach is an intentional simplification compared to the PV/PVC
+model for volumes because we don't need to deal with two objects when
+allocating resources and therefore don't need something like the
+volume binding controller. If desired, a resource plugin can implement
+support for manually allocated (“static provisioning” in the context
+of volumes) and/or existing resources by reacting to ResourceClaims by
+using those resources to satisfy a claim.
+
+Allocation of a resource happens either immediately (“immediate
+allocation”) or when a Pod needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec. Pods reference resource
+claims in a new PodSpec.Resources list. Each resource in that list
+then can be made available to one or more containers in that Pod.  To
+support ephemeral resources, an entry in the new PodSpec.Resources
+list can also be a ResourceClaimTemplate. When a Pod gets created,
+such a template will be used to create a normal ResourceClaim with the
+Pod as owner, and then the normal allocation of the resource takes
+place.
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. For delayed allocation, a node is selected tentatively
```

Comment:
The delayed allocation part is going to be tricky, but I think not impossible, to handle in Cluster Autoscaler. It seems the decision is already made at the Filter stage and only reversed in case of an error, which I think is good enough for autoscaler case. 
However, the details on how we implement it are important here - for example, the scheduler would somehow need to communicate the tentative node selection to Cluster Autoscaler. Can we re-use NominatedNodeName for that? Or otherwise mark it in the pod status?

Reply:
I agree that this is the tricky part. The recent change in the KEP from an explicit allow list for nodes to a deny list was partially motivated by Cluster Autoscaler: this will enable the Filter phase to proceed with a positive response for a fictional node. This may lead to creating nodes that then turn out to be not usable, but if that only happens once per node group it may be okay. We'll just have to figure out how to avoid oscillating between having and not having such a node.

Let's move forward with an alpha implementation that actually allows us to try this out :sweat_smile:

Reply:
Agreed, and added.

Reply:
Since then, a new section with a tenative plan for autoscaler support has been added. I haven't prototyped that, but I think it should work.